src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+
+        /// <summary>
+        /// Combined enqueue/dequeue operation, generally more efficient than sequential Enqueue/Dequeue calls.
+        /// </summary>
+        public TElement EnqueueDequeue(TElement element, TPriority priority)
+        {
+            if (element is null)
+            {
+                throw new ArgumentNullException(nameof(element));
+            }
+
+            if (priority is null)
+            {
+                throw new ArgumentNullException(nameof(priority));
+            }
+
+            var root = _nodes[RootIndex];
+
+            if (this.Comparer.Compare(priority, root.priority) < 0)
+            {
+                return element;
+            }
+            else
+            {
+                var newRoot = (element, priority);
+                _nodes[RootIndex] = newRoot;
+
+                MoveDown(newRoot, RootIndex);
+                _version++;
+
+                return root.element;
+            }
+        }
+
+        /// <summary>
+        /// Enqueues a collection of element/priority pairs.
+        /// </summary>
+        public void EnqueueRange(IEnumerable<(TElement Element, TPriority Priority)> items)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            foreach (var (element, priority) in items)
+            {
+                Enqueue(element, priority);
+            }
+        }
+
+        /// <summary>
+        /// Enqueues a collection of elements, each associated with the specified priority.
+        /// </summary>
+        public void EnqueueRange(IEnumerable<TElement> elements, TPriority priority)
+        {
+            if (elements is null)
+            {
+                throw new ArgumentNullException(nameof(elements));
+            }
+
+            foreach (var element in elements)
+            {
+                Enqueue(element, priority);
+            }
+        }
+
+        /// <summary>
+        /// Removes all items from the priority queue.
+        /// </summary>
+        public void Clear()
+        {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<(TElement, TPriority)>())
+            {
+                // Clear the elements so that the gc can reclaim the references
+                Array.Clear(_nodes, 0, _size);
+            }
+            _size = 0;
+            _version++;
+        }
+
+        /// <summary>
+        /// Ensures that the priority queue has the specified capacity
+        /// and resizes its underlying array if necessary.
+        /// </summary>
+        public void EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(capacity), capacity, SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            if (capacity <= _size)
+            {
+                return;
+            }
+
+            SetCapacity(capacity);
```

Comment:
Note that most `EnsureCapacity` implementations do not set the underlying buffer to exactly match the specified paramater (since the next enqueue would result in a second resize). The [`List<T>` implementation](https://github.com/dotnet/runtime/blob/dfc70d174ff13894d9ca6added89601394c0ea2a/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L399) is probably a good template to follow here.

Reply:
Right, this is in contrast to the TrimExcess pattern (compare both on `List<T>` or on [`Dictionary<K,V>`](https://source.dot.net/#System.Private.CoreLib/Dictionary.cs,1070))